client/src/cmdhffelica.c

```diff
@@ -2247,6 +2251,430 @@ static int CmdHFFelicaDumpLite(const char *Cmd) {
     return PM3_SUCCESS;
 }
 
+static void reverse_byte_array(const uint8_t* in, const int length, uint8_t* out) {
+    for (int i = 0; i < length; i++) {
+        out[i] = in[(length - 1) - i];
+    }
+}
+
+static void reverse_block16(const uint8_t* in, uint8_t *out) {
+    reverse_byte_array(in, 8, out);
+    reverse_byte_array(in + 8, 8, out + 8);
+}
+
+static int felica_make_block_list(uint16_t *out, const uint8_t *blk_numbers, const size_t length) {
+    if (length > 4) {
+        PrintAndLogEx(ERR, "felica_make_block_list: exceeds max size");
+        return PM3_EINVARG;
+    }
+
+    uint16_t tmp[4];
+    memset(tmp, 0, sizeof(tmp));
+
+    for (size_t i = 0; i < length; i++) {
+        tmp[i] = (uint16_t)(blk_numbers[i] << 8) | 0x80;
+    }
+    memcpy(out, tmp, length * sizeof(uint16_t));
+
+    return PM3_SUCCESS;
+}
+
+static int read_without_encryption(
+    const uint8_t *idm,
+    const uint8_t num,
+    const uint8_t *blk_numbers,
+    uint8_t *out,
+    uint16_t *n) {
+
+    felica_read_request_haeder_t request = {
+        .command_code = { 0x06 },
+        .number_of_service = { 1 },
+        .service_code_list = { 0x00B },
+        .number_of_block = { num },
+    };
+    memcpy(request.IDm, idm, sizeof(request.IDm));
+
+    uint16_t svc[num];
+    int ret = felica_make_block_list(svc, blk_numbers, num);
+    if (ret) {
+        return PM3_EINVARG;
+    }
+
+    size_t hdr_size = sizeof(request);
+    uint16_t size = hdr_size + sizeof(svc) + 1;
+    *n = size;
+
+    memcpy(out, &(uint8_t){ size }, sizeof(uint8_t));
+    memcpy(out + 1, &request, hdr_size);
+    memcpy(out + hdr_size + 1, &svc, sizeof(svc));
+
+    return PM3_SUCCESS;
+}
+
+static bool check_write_req_data(const felica_write_request_haeder_t *hdr, const uint8_t datalen) {
+    if (!hdr || !datalen)
+        return false;
+
+    uint8_t num = *(hdr->number_of_block);
+    if (num != 1 && num != 2)
+        return false;
+    
+    // Check Block data size
+    if (num * 16 != datalen)
+        return false;
+
+    return true;
+}
+
+static int write_without_encryption( 
+    const uint8_t *idm, 
+    uint8_t num, 
+    uint8_t *blk_numbers, 
+    const uint8_t *data, 
+    size_t datalen,
+    uint8_t *out,
+    uint16_t *n) {
+
+    felica_write_request_haeder_t hdr = {
+        .command_code = { 0x08 },
+        .number_of_service = { 1 },
+        .service_code_list = { 0x009 },
+        .number_of_block = { num },
+    };
+    memcpy(hdr.IDm, idm, sizeof(hdr.IDm));
+
+    uint8_t dl = (uint8_t)(datalen);
+
+    if (!check_write_req_data(&hdr, dl)) {
```

Comment:
Code style: we use  ` <expression> == false`    and not `!<expression>`     - since it helps readability.